src/Umbraco.Infrastructure/Migrations/Upgrade/V_15_0_0/ConvertLocalLinks.cs

```diff
@@ -0,0 +1,295 @@
+    {
+        switch (editorValue)
+        {
+            case RichTextEditorValue richTextValue:
+                return ProcessRichText(richTextValue);
+            case BlockValue blockValue:
+                return ProcessBlocks(blockValue);
+        }
+
+        return false;
+    }
+
+    private bool ProcessRichText(RichTextEditorValue richTextValue)
+    {
+        bool hasChanged = false;
+
+        var newMarkup = ProcessStringValue(richTextValue.Markup);
+        if (newMarkup.Equals(richTextValue.Markup) == false)
+        {
+            hasChanged = true;
+            richTextValue.Markup = newMarkup;
+        }
+
+        if (richTextValue.Blocks is null)
+        {
+            return hasChanged;
+        }
+
+        foreach (BlockItemData blockItemData in richTextValue.Blocks.ContentData)
+        {
+            foreach (BlockPropertyValue blockPropertyValue in blockItemData.Values)
+            {
+                if (ProcessToEditorValue(blockPropertyValue.Value))
+                {
+                    hasChanged = true;
+                }
+            }
+        }
+
+        return hasChanged;
+    }
+
+    private bool ProcessBlocks(BlockValue blockValue)
+    {
+        bool hasChanged = false;
+
+        foreach (BlockItemData blockItemData in blockValue.ContentData)
+        {
+            foreach (BlockPropertyValue blockPropertyValue in blockItemData.Values)
+            {
+                if (ProcessToEditorValue(blockPropertyValue.Value))
+                {
+                    hasChanged = true;
+                }
+            }
+        }
+
+        return hasChanged;
+    }
+
+    private string ProcessStringValue(string input)
+    {
+        // find all legacy tags
+        IEnumerable<HtmlLocalLinkParser.LocalLinkTag> tags = _localLinkParser.FindLegacyLocalLinkIds(input);
+
+        foreach (HtmlLocalLinkParser.LocalLinkTag tag in tags)
+        {
+            string newTagHref;
+            if (tag.Udi is not null)
+            {
+                newTagHref = $" type=\"{tag.Udi.EntityType}\" "
+                             + tag.TagHref.Replace(tag.Udi.ToString(), tag.Udi.Guid.ToString());
+            }
+            else if (tag.IntId is not null)
+            {
+                // try to get the key and type from the int, else do nothing
+                (Guid Key, string EntityType)? conversionResult = CreateIntBasedTag(tag.IntId.Value);
+                if (conversionResult is null)
+                {
+                    continue;
+                }
+
+                newTagHref = $" type=\"{conversionResult.Value.EntityType}\" "
+                             + tag.TagHref.Replace(tag.IntId.Value.ToString(), conversionResult.Value.Key.ToString());
+            }
+            else
+            {
+                // tag does not contain enough information to convert
+                continue;
+            }
+
+            input = input.Replace(tag.TagHref, newTagHref);
```

Comment:
Note that the existing `href` tag might contain hashes and querystrings that are not part of the `tag.TagHref`, e.g. `/some-url#anchor` or `/some-other-url?v=1`. The resulting (new) href should look like this in that case `/{localLink:<GUID>}#anchor` and `/{localLink<GUID>}?v=1` respectively.

Reply:
If I recall correctly, the tag.TagHref _should_ avoid those, so you are replacing only the UDI part of the link, so you are probably safe, but wanted to leave this comment here just in case :-)

Reply:
I will double check this, thx Jacob!

Reply:
The old href detection only returns `href="{localLink:umb://document/<ID>}`
Inside that string I am only replacing either the full udi `umb://document/<ID>` or the old old int value by the guid and then prepending the string with `type="<NodeType>"` so it is indeed safe.